les/clientpool.go

```diff
@@ -103,296 +106,180 @@ type clientPool struct {
-func connPriority(a interface{}, now mclock.AbsTime) int64 {
-	c := a.(*clientInfo)
-	return c.balanceTracker.getPriority(now)
-}
-
-// connMaxPriority callback returns estimated maximum priority of clientInfo item in connectedQueue
-func connMaxPriority(a interface{}, until mclock.AbsTime) int64 {
-	c := a.(*clientInfo)
-	pri := c.balanceTracker.estimatedPriority(until, true)
-	c.balanceTracker.addCallback(balanceCallbackQueue, pri+1, func() {
-		c.pool.lock.Lock()
-		if c.queueIndex != -1 {
-			c.pool.connectedQueue.Update(c.queueIndex)
+// newClientPool creates a new client pool
+func newClientPool(lespayDb ethdb.Database, minCap, freeClientCap uint64, activeBias time.Duration, clock mclock.Clock, removePeer func(enode.ID)) *clientPool {
+	if minCap > freeClientCap {
+		panic(nil)
+	}
+	ns := nodestate.NewNodeStateMachine(nil, nil, clock, clientPoolSetup)
+	pool := &clientPool{
+		ns:                  ns,
+		BalanceTrackerSetup: balanceTrackerSetup,
+		PriorityPoolSetup:   priorityPoolSetup,
+		clock:               clock,
+		minCap:              minCap,
+		freeClientCap:       freeClientCap,
+		removePeer:          removePeer,
+	}
+	pool.ti = lps.NewTokenIssuer(ns, tokenIssuerSetup, clock, freeClientCap, activeBias/8, time.Hour)
+	pool.posExp = pool.ti.NewTokenExpirer()
+	pool.negExp = pool.ti.NewTokenExpirer()
+	pool.bt = lps.NewBalanceTracker(ns, balanceTrackerSetup, lespayDb, clock, pool.posExp, pool.negExp)
+	pool.ti.SetTotalAmountCallback(pool.bt.TotalTokenAmount)
+	pool.pp = lps.NewPriorityPool(ns, priorityPoolSetup, clock, minCap, activeBias, 4)
+
+	// set default expiration constants used by tests
+	// Note: server overwrites this if token sale is active
+	pool.bt.SetExpirationTCs(0, defaultNegExpTC)
+	// calculate total token balance amount
+
+	ns.SubscribeState(pool.InactiveFlag.Or(pool.PriorityFlag), func(node *enode.Node, oldState, newState nodestate.Flags) {
+		if newState.Equals(pool.InactiveFlag) {
+			ns.AddTimeout(node, pool.InactiveFlag, inactiveTimeout)
+		}
+		if oldState.Equals(pool.InactiveFlag) && newState.Equals(pool.InactiveFlag.Or(pool.PriorityFlag)) {
+			ns.SetState(node, pool.InactiveFlag, nodestate.Flags{}, 0) // remove timeout
 		}
-		c.pool.lock.Unlock()
 	})
-	return pri
-}
 
-// priceFactors determine the pricing policy (may apply either to positive or
-// negative balances which may have different factors).
-// - timeFactor is cost unit per nanosecond of connection time
-// - capacityFactor is cost unit per nanosecond of connection time per 1000000 capacity
-// - requestFactor is cost unit per request "realCost" unit
-type priceFactors struct {
-	timeFactor, capacityFactor, requestFactor float64
-}
+	ns.SubscribeState(pool.ActiveFlag.Or(pool.PriorityFlag), func(node *enode.Node, oldState, newState nodestate.Flags) {
+		if newState.Equals(pool.ActiveFlag) {
+			cap, _ := ns.GetField(node, pool.CapacityField).(uint64)
+			if cap > freeClientCap {
+				pool.pp.RequestCapacity(node, freeClientCap, 0, true)
```

Comment:
If i understand correctly, the logic here is if the capacity exceeds the freecap, reset it back? Why

Reply:
This happens when the node is active but has no priority status (zero positive balance). If a paid client had a high capacity but the positive balance has expired then it becomes a free client and the capacity is reduced to free cap.